vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,200 @@
+
+Although git can store binary files, its real power and efficiency is with plain text files. R has several functions to read and write `data.frames` as plain text files. Main downside is the loss of meta data, especially in case of `factors`. A `factor` will be written as a `character` and hence lose all information on the levels. This can be important when the levels are a) not sorted alphabetical, b) not all levels have observations or c) the `factor` is ordered. Besides that, storing a `factor` as `character` is not efficient.
+
+Git stores the version history under the form of diffs: a list of lines which are deleted and a list of lines which are inserted at a specific line number in a file. We have to keep this in mind when storing data as plain text files in Git. The table below indicate what impact changing the information content has on the diff in Git. Note that changing observations requires a small diff, while changing variables results in a massive diff.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("remove 1 observation", "remove 1 line"),
+    c("add 1 observation", "add 1 line"),
+    c("update 1 observation", "remove 1 line and add 1 line"),
+    c("remove 1 variable", "remove all lines and add all lines"),
+    c("add 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the information content of the data")
+```
+
+In the next table we show the effect on the diff when making changes in the data which don't change in the information content in the data.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("move 1 observation", "remove 1 line and add 1 line"),
+    c("move 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the data without changing the information content")
+```
+
+So in order to use Git a a performant and efficient version control system for data, we need to make sure to store the meta data and keep the diffs as small as possible.
+
+## Tools in `git2r`
+
+### Data repository
+
+A 'data repository' refers to a sub directory within the root of an existing git repository. Create a `data_repository` object with `init()` or `repository()` and specify the sub directory through the `project` argument. Most functions of the package handle a `data_repository` object in the same way as they handle a `git_repository` object. The main difference is that the working directory points to the sub directory of the root, rather that the root of the git repository. You can create multiple data repositories within the same git repository.
```

Comment:
As stated in the `repository.R` file: I would not use the name `data-repository` with files belonging to a data-repository and multiple data-repositories belonging to a git repository... I think this is hampers the usage for new users.  I assume it is convenient with the usage of S3 object to provide specialized merge/... methods for data-repositories, but I really do think this hampers ease of use. 

What I would like - as a user - is to not have to bother about data-repositories, but just to be able to use the `read_delim_git()` and  `write_delim_git()` functions within a git-repo to read and write version-controlled data files, as these functions (while remaining factor-information,...) are providing the added value.